crates/db_connection_pool/src/clickhousepool.rs

```diff
@@ -109,6 +92,91 @@ impl ClickhouseConnectionPool {
     }
 }
 
+const DEFAULT_CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+
+fn get_config_from_params(
+    params: &HashMap<String, String>,
+    secret: &Option<Secret>,
+) -> Result<Options> {
+    let options: Option<Options>;
+    if let Some(clickhouse_connection_string) = get_secret_or_param(
+        params,
+        secret,
+        "clickhouse_connection_string_key",
+        "clickhouse_connection_string",
+    ) {
+        let mut new_options = Options::from_str(&clickhouse_connection_string)
+            .context(InvalidConnectionStringSnafu)?;
+        if !clickhouse_connection_string.contains("connection_timeout") {
+            // Default timeout of 500ms is not enough in some cases.
+            new_options = new_options.connection_timeout(DEFAULT_CONNECTION_TIMEOUT);
+        }
+        options = Some(new_options);
+    } else {
+        let mut connection_opts: Vec<String> = vec![];
+        let keys = [
+            "clickhouse_user",
+            "clickhouse_host",
+            "clickhouse_tcp_port",
+            "clickhouse_db",
+        ];
+        for key in &keys {
+            let value = params.get(*key);
+            match value {
+                Some(value) => connection_opts.push(value.clone()),
+                None => NotEnoughParametersForConnectionSnafu {
+                    parameter_name: (*key).to_string(),
+                }
+                .fail()?,
+            }
+        }
+        let password =
+            get_secret_or_param(params, secret, "clickhouse_pass_key", "clickhouse_pass");
+        match password {
+            Some(password) => connection_opts.insert(1, password),
+            None => NotEnoughParametersForConnectionSnafu {
+                parameter_name: "clickhouse_pass".to_string(),
+            }
+            .fail()?,
+        }
+        let connection_string = format!(
+            "tcp://{}:{}@{}:{}/{}",
+            connection_opts[0],
+            connection_opts[1],
+            connection_opts[2],
+            connection_opts[3],
+            connection_opts[4]
+        );
+        // Default timeout of 500ms is not enough
+        let new_options = Options::from_str(&connection_string)
+            .context(InvalidConnectionStringSnafu)?
+            .connection_timeout(DEFAULT_CONNECTION_TIMEOUT);
+        options = Some(new_options);
+    }
+    let mut options = options.ok_or(Error::NotEnoughParametersForConnection {
+        parameter_name: "clickhouse_connection_string".to_string(),
+    })?;
+    let secure = params
+        .get("clickhouse_secure")
+        .map(|s| s.parse::<bool>())
+        .transpose()
+        .context(InvalidSecureParameterValueSnafu {
+            parameter_name: "clickhouse_secure".to_string(),
+        })?;
+    options = options.secure(secure.unwrap_or(true));
+
+    if let Some(clickhouse_sslrootcert) = params.get("clickhouse_sslrootcert") {
```

Comment:
Where is `clickhouse_sslrootcert` used? Should it be passed to options?

Reply:
Removed, currently there is a problem this crate.
It doesn't expose the required code.